test/release/examples/primers/ranges.chpl

```diff
@@ -7,245 +7,384 @@
   sum += i;
-writeln("The sum of the values in ", r, " is ", sum);
-
-//
-// We can compute this same sum with a reduction over the range.
-//
-sum = + reduce r;
-writeln("The sum of the values in ", r, " as computed by reduce is also ", sum);
+writeln("The sum of the values in '1..10' is ", sum);
 writeln();
 
+// Ranges can also drive parallel loop forms such as
+// :ref:`forall<primers-forallLoops>` and
+// :ref:`coforall<primers-taskparallel-coforall>` loops.
+
 //
-// Ranges are also the building blocks for domains.  Here
-// we use the range ``r`` to build up a 2-D domain and an array.
+// Ranges also serve as the building block for rectangular
+// :ref:`domains <primers-domains>` and :ref:`arrays
+// <primers-arrays>`.  Here we use the range ``r`` to declare a 2D
+// 10x10 domain and an array over it.
 //
 writeln("Domains and arrays");
-const D: domain(2) = {r, r};
-var A: [D] real;
+const D = {1..10, 1..10};  // could also use '= {r, r};'
+var A: [D] real;           // a 10x10 array of real floating point values
 writeln("D = ", D);
 writeln("Array A");
 writeln(A);
 writeln();
 
+// Arrays can also be declared in terms of anonymous domains by using
+// ranges directly (optionally dropping the curly brackets):
+
+var A1: [1..n] int,
+    A2: [1..n, 1..n] string;
+
 //
-// Ranges don't have to be fully bounded.  A range without a low bound
-// represents an infinite sequence of numbers that are less than or equal to
-// the high bound.  Similarly a range without a high bound represents
-// an infinitude of numbers greater or equal to the low bound.  A range
-// with no bounds is allowed as well.
+// Ranges can be used in general contexts that expect an iterable
+// collection of values, such as reductions:
 //
-writeln("Unbounded ranges");
-writeRange(1..);
-writeRange(..5);
-writeRange(..);
+sum = + reduce r;  // compute 1 + 2 + 3 + ... + 10
+writeln("The sum of the values in ", r, ", computed by reduce, is also ", sum);
+
+// Or whole-array operations:
+A1 = r;      // assign the elements of A1 their corresponding values of r
+writeln(A1);
+writeln();
 
+
+/*
+  Variations on Basic Ranges
+  --------------------------
+*/
+
+// Ranges can be defined over boolean values and enumerated types.  In
+// the case of enumerations, the range's values are ordered according
+// to the declaration order of the enum's constants rather than their
+// (optional) numerical values.
+
+writeln("Ranges over bools and enums:");
+enum dir {north, south, east, west};
+enum color {red=4, orange=2, yellow=1, green=3, blue=6, indigo=7, violet=5};
+const boolRange = false..true,                 // false, true
+      enumRange = dir.north..dir.west,         // north, south, east, west
+      colorRange = color.orange..color.green;  // orange, yellow, green
+writeRange(boolRange);
+writeRange(enumRange);
+writeRange(colorRange);
+writeln();
+
+//
+// Ranges can also be `unbounded` on one or both sides, when the
```

Comment:
I've moved the unbounded part ahead of bool/enum ranges if that satisfies.  I didn't move them higher mostly because I think they're used most frequently in zippered contexts, which we haven't mentioned before now and because I think most of the previous sections are more important (except for maybe the "ranges can be used like they were collections" part, though there's not really a better place for it).

Also: bounded enum ranges rock!!  You don't know what you're missing!  :D

Reply:
Yeah I think that's good. Especially with the rewording in the earlier parts making it clear that ranges can be unbounded.